src_lockfree/htbl/htbl.body.ml

```diff
@@ -0,0 +1,617 @@
+  && begin
+       finish t new_r |> ignore;
+       true
+     end
+
+let adjust_size t r mask result =
+  if
+    r.pending == Nothing
+    && Int64.to_int (Random.bits64 ()) land mask = 0
+    && Atomic.get t == r
+  then begin
+    (* Reading the size is potentially expensive, so we only check it
+       occasionally.  The bigger the table the less frequently we should need to
+       resize. *)
+    let size = Size.get r.size in
+    let capacity = Atomic_array.length r.buckets in
+    if capacity < size && capacity < r.max_buckets then
+      try_resize t r (capacity + capacity) ~clear:false |> ignore
+    else if r.min_buckets < capacity && size + size + size < capacity then
+      try_resize t r (capacity lsr 1) ~clear:false |> ignore
+  end;
+  result
+
+(* *)
+
+(** [get] only returns with a state where [pending = Nothing]. *)
+let[@inline] get t =
+  let r = Atomic.get t in
+  if r.pending == Nothing then r else finish t r
+
+(* *)
+
+let rec assoc_node t key = function
+  | Nil -> (Nil : (_, _, [< `Nil | `Cons ]) tdt)
+  | Cons r as cons -> if t r.key key then cons else assoc_node t key r.rest
+
+let rec assoc r key = function
+  | B (Nil_with_size nil_r) ->
+      if nil_r.size_modifier != Size.used_once then
+        Size.update_once r.size nil_r.size_modifier;
+      Nil
+  | B (Cons_with_size cons_r) ->
+      if cons_r.size_modifier != Size.used_once then
+        Size.update_once r.size cons_r.size_modifier;
+      if r.equal cons_r.key key then
+        Cons { key = cons_r.key; value = cons_r.value; rest = cons_r.rest }
+      else assoc_node r.equal key cons_r.rest
+  | B (Resize resize_r) ->
+      (* A resize is in progress.  The spine of the resize still holds what was
+         in the bucket before resize reached that bucket. *)
+      assoc r key (B resize_r.spine)
+
+let find_node t key =
+  (* Reads can proceed in parallel with writes. *)
+  let r = Atomic.get t in
+  let h = r.hash key in
+  let mask = Atomic_array.length r.buckets - 1 in
+  let i = h land mask in
+  assoc r key @@ Atomic_array.unsafe_fenceless_get r.buckets i
+(* *)
+
+let find_exn t key =
+  match find_node t key with
+  | Nil -> raise_notrace Not_found
+  | Cons r -> r.value
+
+let mem t key = find_node t key != Nil
+
+(* *)
+
+let rec try_add t key value backoff =
+  let r = Atomic.get t in
+  let h = r.hash key in
+  let mask = Atomic_array.length r.buckets - 1 in
+  let i = h land mask in
+  match Atomic_array.unsafe_fenceless_get r.buckets i with
+  | B (Nil_with_size nil_r as before) ->
+      if nil_r.size_modifier != Size.used_once then
+        Size.update_once r.size nil_r.size_modifier;
+      let (Cons_with_size cons_r as after : ('a, 'b, [ `Cons_with_size ]) tdt) =
+        Cons_with_size
+          {
+            size_modifier = Size.new_once r.size Size.incr;
+            key;
+            value;
+            rest = Nil;
+          }
+      in
+      if Atomic_array.unsafe_compare_and_set r.buckets i (B before) (B after)
+      then (
```

Comment:
I'd prefer if we used `begin ... end` for blocks with semicolon separated statements.  I find them much easier to read and I like that the `begin ... end` alerts me that there are likely side-effects.  There are a few other places with `( ... )` blocks.